tests/test_unittest/test_storage.py

```diff
@@ -318,68 +318,92 @@ def test_allocate_gpu_unmanaged(param_dict):
+    assert shape_out == shape
+    assert dtype_out == dtype
+    assert mask_out == mask
+
+    # Default origin
+    default_origin_out, shape_out, dtype_out, mask_out = normalize_storage_spec(
+        (1, 1, 1), shape, dtype, mask
+    )
+    assert default_origin_out == (1, 1, 1)
+    assert shape_out == shape
+    assert dtype_out == dtype
+    assert mask_out == mask
+
+    with pytest.raises(TypeError, match="default_origin"):
+        normalize_storage_spec(None, shape, dtype, mask)
+    with pytest.raises(TypeError, match="default_origin"):
+        normalize_storage_spec(("1", "1", "1"), shape, dtype, mask)
+    with pytest.raises(ValueError, match="default_origin"):
+        normalize_storage_spec((1, 1, 1, 1), shape, dtype, mask)
+    with pytest.raises(ValueError, match="default_origin"):
+        normalize_storage_spec((-1, -1, -1), shape, dtype, mask)
+
+    # Shape
+    default_origin_out, shape_out, dtype_out, mask_out = normalize_storage_spec(
+        default_origin, (10, 10), dtype, (True, True, False)
+    )
+    assert default_origin_out == (0, 0)
+    assert shape_out == (10, 10)
+    assert dtype_out == dtype
+    assert mask_out == (True, True, False)
+
+    _, shape_out, __, mask_out = normalize_storage_spec(
+        default_origin, (10, 20), dtype, (False, True, False)
+    )
+    assert shape_out == (20,)
+    assert mask_out == (False, True, False)
+
+    with pytest.raises(TypeError, match="shape"):
+        normalize_storage_spec(default_origin, "(10,10)", dtype, mask)
+    with pytest.raises(TypeError, match="shape"):
+        normalize_storage_spec(default_origin, None, dtype, mask)
+    with pytest.raises(ValueError, match="shape"):
+        normalize_storage_spec(default_origin, (10, 10, 0), dtype, mask)
+
+    # Mask
+    default_origin_out, shape_out, dtype_out, mask_out = normalize_storage_spec(
+        default_origin, (10, 10), dtype, (True, True, False)
+    )
+    assert default_origin_out == (0, 0)
+    assert shape_out == (10, 10)
+    assert dtype_out == dtype
+    assert mask_out == (True, True, False)
+
+    _, __, ___, mask_out = normalize_storage_spec(default_origin, (10, 10), dtype, "IJ")
+    assert mask_out == (True, True, False)
+
+    _, __, ___, mask_out = normalize_storage_spec(default_origin, (10, 10), dtype, gtscript.IJ)
+    assert mask_out == (True, True, False)
+
+    _, __, ___, mask_out = normalize_storage_spec(default_origin, (10, 10, 10), dtype, gtscript.IJK)
+    assert mask_out == (True, True, True)
+
+    with pytest.raises(ValueError, match="mask"):
+        normalize_storage_spec(default_origin, (10, 10), dtype, (False, False, False))
+
+    # Dtype
+    default_origin_out, shape_out, dtype_out, mask_out = normalize_storage_spec(
+        default_origin, shape, (dtype, (2,)), mask
+    )
+    assert default_origin_out == tuple([*default_origin, 0])
+    assert shape_out == tuple([*shape, 2])
+    assert dtype_out == dtype
+    assert mask_out == tuple([*mask, True])
```

Comment:
Github doesn't let me put this comment on the `test_masked_storage_?pu` below because it's not part of the commit, but there, the skip should now be removed again and the test fixed.